src/orion/client/runner.py

```diff
@@ -0,0 +1,323 @@
+
+        self.interrupt_signal_code = interrupt_signal_code
+
+    @property
+    def is_done(self):
+        """Returns true if the experiment has finished."""
+        return self.client.is_done
+
+    @property
+    def has_remaining(self):
+        """Returns true if the worker can still pick up work"""
+        return (
+            self.max_trials_per_worker - (self.trials - self.worker_broken_trials) > 0
+        )
+
+    @property
+    def is_idle(self):
+        """Returns true if none of the workers are running a trial"""
+        return len(self.pending_trials) <= 0
+
+    @property
+    def running(self):
+        """Returns true if we are still running trials."""
+        return self.pending_trials or (self.has_remaining and not self.is_done)
+
+    def run(self):
+        """Run the optimizing process until completion.
+
+        Returns
+        -------
+        the total number of trials processed
+
+        """
+        idle_start = time.time()
+        idle_end = 0
+        idle_time = 0
+
+        while self.running:
+
+            # Get new trials for our free workers
+            with self.stat.time("sample"):
+                new_trials = self.sample()
+
+            # Scatter the new trials to our free workers
+            with self.stat.time("scatter"):
+                self.scatter(new_trials)
+
+            # Gather the results of the workers that have finished
+            with self.stat.time("gather"):
+                self.gather()
+
+            if self.is_idle:
+                idle_end = time.time()
+                idle_time += idle_end - idle_start
+                idle_start = idle_end
+
+                log.debug(f"Workers have been idle for {idle_time:.2f} s")
+            else:
+                idle_start = time.time()
+                idle_time = 0
+
+            if self.is_idle and idle_time > self.idle_timeout:
+                msg = f"Workers have been idle for {idle_time:.2f} s"
+
+                if self.pending_trials:
+                    msg = (
+                        f"{msg}; but not all trials finished; "
+                        f"pending_trials: {self.pending_trials}"
+                    )
+
+                if self.has_remaining and not self.is_done:
+                    msg = (
+                        f"{msg}; worker has leg room (has_remaining: {self.has_remaining})"
+                        f" and optimization is not done (is_done: {self.is_done})"
+                    )
+
+                raise LazyWorkers(msg)
+
+        return self.trials
+
+    def sample(self):
+        """Sample new trials for all free workers"""
+        ntrials = len(self.pending_trials) + self.trials
+        remains = self.max_trials_per_worker - ntrials
+
+        # try to get more work
+        new_trials = []
+        n_trial = min(self.free_worker, remains)
+        should_sample_more = (not self.is_done) and self.free_worker > 0 and remains > 0
+
+        if should_sample_more:
```

Comment:
The logic should be tested here. Is `n_trial` passed to `self._suggest_trials()` the value we expect, in a case where we have no pending_trials, no free worker, less remaining than free workers or vice versa, if the experiment is done or broken. There are many moving parts.

Reply:
added